aws/session/session.go

```diff
@@ -21,36 +15,128 @@ import (
+// the shared config file (~/.aws/config) will also be loaded in addition to
+// the shared credentials file (~/.aws/config). Values set in both the shared
+// shared config, and shared credentials will be taken from the shared
+// credentials file. Enabling the Shared Config will also allow the Session
+// to be built with retrieving credentials with AssumeRole set in the config.
+func Build(cfgs ...*aws.Config) (*Session, error) {
+	envcfg := loadEnvConfig()
+
+	return buildSession(envcfg, cfgs...)
+}
+
+// BuildFromProfile returns a Session built from SDK defaults, config files,
+// environment, and user provided config files.
 //
-//     // Create a session with a custom region
-//     sess := session.New(&aws.Config{Region: aws.String("us-east-1")})
+// Uses the profile to configure which profile to load the configuration from.
+// This will override the profile specified in the environment.
 //
-//     // Create a session, and add additional handlers for all service
-//     // clients created with the session to inherit. Adds logging handler.
-//     sess := session.New()
-//     sess.Handlers.Send.PushFront(func(r *request.Request) {
-//          // Log every request made and its payload
-//          logger.Println("Request: %s/%s, Payload: %s", r.ClientInfo.ServiceName, r.Operation, r.Params)
-//     })
+// Same as Build, but specifies the profile that will be loaded from the shared
+// configuration files. This function also reads the environment variable
+// AWS_SDK_LOAD_CONFIG the same as Build.
+func BuildFromProfile(profile string, cfgs ...*aws.Config) (*Session, error) {
+	envcfg := loadEnvConfig()
+	envcfg.Profile = profile
+
+	return buildSession(envcfg, cfgs...)
+}
+
+// BuildFromSharedConfig returns a Session built from SDK defaults, config
+// files, environment, and user provided config files.
 //
-//     // Create a S3 client instance from a session
-//     sess := session.New()
-//     svc := s3.New(sess)
-func New(cfgs ...*aws.Config) *Session {
+// Same as Build, except it builds the Session as if the shared config was
+// enabled e.g. AWS_SDK_LOAD_CONFIG=1.
+func BuildFromSharedConfig(cfgs ...*aws.Config) (*Session, error) {
+	envCfg := loadSharedEnvConfig()
+
+	return buildSession(envCfg, cfgs...)
+}
+
+// BuildFromSharedConfigProfile returns a Session built from SDK defaults,
+// config files, environment, and user provided config files.
+//
+// Uses the profile to configure which profile to load the shared configuration
+// from. This will override the profile specified in the environment.
+//
+// Same as Build, except it builds the Session as if the shared config was
+// enabled e.g. AWS_SDK_LOAD_CONFIG=1.
+func BuildFromSharedConfigProfile(profile string, cfgs ...*aws.Config) (*Session, error) {
+	// Load initial config from environment
+	envCfg := loadSharedEnvConfig()
+	envCfg.Profile = profile
+
+	return buildSession(envCfg, cfgs...)
+}
+
+func oldNewSession(cfgs ...*aws.Config) *Session {
```

Comment:
I suggest renaming this to something more obvious like ` newSessionFallback ` or similar.